src/algebra/free_algebra.lean

```diff
@@ -0,0 +1,293 @@
+      rw mul_assoc },
+    { change (algebra_map _ _ _) * lift_fun R X f _ = lift_fun R X f _,
+      simp, },
+    { change lift_fun R X f _ * (algebra_map _ _ _) = lift_fun R X f _,
+      simp, },
+    { change _ * (_ + _) = _ * _ + _ * _,
+      rw left_distrib, },
+    { change (_ + _) * _ = _ * _ + _ * _,
+      rw right_distrib, },
+    { change (algebra_map _ _ _) * _ = algebra_map _ _ _,
+      simp },
+    { change _ * (algebra_map _ _ _) = algebra_map _ _ _,
+      simp },
+    repeat { change lift_fun R X f _ + lift_fun R X f _ = _,
+      rw h_ih,
+      refl, },
+    repeat { change lift_fun R X f _ * lift_fun R X f _ = _,
+      rw h_ih,
+      refl, },
+  end,
+  map_one' := by { change algebra_map _ _ _ = _, simp },
+  map_mul' := by { rintros ⟨⟩ ⟨⟩, refl },
+  map_zero' := by { change algebra_map _ _ _ = _, simp },
+  map_add' := by { rintros ⟨⟩ ⟨⟩, refl },
+  commutes' := by tauto }
+
+variables {R X}
+
+@[simp]
+theorem ι_comp_lift {A : Type*} [semiring A] [algebra R A] (f : X → A) :
+  (lift R X f : free_algebra R X → A) ∘ (ι R X) = f := by {ext, refl}
+
+@[simp]
+theorem lift_ι_apply {A : Type*} [semiring A] [algebra R A] (f : X → A) (x) :
+  lift R X f (ι R X x) = f x := rfl
+
+@[simp]
+theorem lift_unique {A : Type*} [semiring A] [algebra R A] (f : X → A)
+  (g : free_algebra R X →ₐ[R] A) :
+  (g : free_algebra R X → A) ∘ (ι R X) = f ↔ g = lift R X f :=
+begin
+  refine ⟨λ hyp, _, λ hyp, by rw [hyp, ι_comp_lift]⟩,
+  ext,
+  rcases x,
+  induction x,
+  { change ((g : free_algebra R X → A) ∘ (ι R X)) _ = _,
+    rw hyp,
+    refl },
+  { exact alg_hom.commutes g x },
+  { change g (quot.mk _ _ + quot.mk _ _) = _,
+    rw [alg_hom.map_add, x_ih_a, x_ih_a_1],
+    refl },
+  { change g (quot.mk _ _ * quot.mk _ _) = _,
+    rw [alg_hom.map_mul, x_ih_a, x_ih_a_1],
+    refl },
+end
+
+attribute [irreducible] free_algebra ι lift
+
+@[simp]
+theorem lift_comp_ι {A : Type*} [semiring A] [algebra R A] (g : free_algebra R X →ₐ[R] A) :
+  lift R X ((g : free_algebra R X → A) ∘ (ι R X)) = g := by {symmetry, rw ←lift_unique}
+
+@[ext]
+theorem hom_ext {A : Type*} [semiring A] [algebra R A] {f g : free_algebra R X →ₐ[R] A}
+  (w : ((f : free_algebra R X → A) ∘ (ι R X)) = ((g : free_algebra R X → A) ∘ (ι R X))) : f = g :=
+begin
+  have : g = lift R X ((g : free_algebra R X → A) ∘ (ι R X)), by rw ←lift_unique,
+  rw [this, ←lift_unique, w],
+end
+
+/--
+The free algebra on `X` is "just" the monoid algebra on the free monoid on `X`.
+
+This would be useful when constructing linear maps out of a free algebra,
+for example.
+-/
+noncomputable
+def equiv_monoid_algebra_free_monoid : free_algebra R X ≃ₐ[R] monoid_algebra R (free_monoid X) :=
+alg_equiv.of_alg_hom
+  (lift R X (λ x, (monoid_algebra.of R (free_monoid X)) (free_monoid.of x)))
+  ((monoid_algebra.lift R (free_monoid X) (free_algebra R X)) (free_monoid.lift (ι R X)))
+begin
+  apply monoid_algebra.alg_hom_ext, intro x,
+  apply free_monoid.rec_on x,
+  { simp, refl, },
```

Comment:
Is `simp, refl` legit?

Reply:
I think we've decided that it is, and I've been quoting https://leanprover.zulipchat.com/#narrow/stream/144837-PR-reviews/topic/.233638/near/209163865 as the canon. :-)

Reply:
I don't know about this example in particular, but `simp, refl` could break if the simp set changed (and I think that's what Mario said we should try to avoid). However, I also don't think it's a big deal, though maybe I'll change my opinion if I ever have to try to fix a bunch of these.

Reply:
This seems quite innocent. `simp; refl` would be slightly more robust, but I don't care.